### ISSUE TITLE
Custom pipeable operator does not unsubscribe source when executed synchronously

### ISSUE THREAD

**Issue by EnricoPicci** [ReactiveX/rxjs#5369]
## Bug Report

**Current Behavior**
I create a custom pipeable operator to filter out null values. Functionally it behaves the same as `filter(v => v !== null)`. It seems though that, when used synchronously, if it is unsubscribed, the custom operator does not propagate the unsubscription to its source.

**Reproduction**
- https://stackblitz.com/edit/rxjs-ksphw9?embed=1&file=index.ts

```
import { of, Observable } from "rxjs";
import { take, tap, filter } from "rxjs/operators";

function filterNil() {
  return function<T>(source: Observable<T>): Observable<T> {
    return new Observable(subscriber => {
      const subscription = source.subscribe({
        next(value) {
          if (value !== undefined && value !== null) {
            subscriber.next(value);
          }
        },
        error(error) {
          subscriber.error(error);
        },
        complete() {
          subscriber.complete();
        }
      });

      return () => subscription.unsubscribe();
    });
  };
}

console.log("With Custom Operator");
of(null, 20, 15, 10, 5, 6, 9)
  .pipe(
    tap(val => console.log(`Tapped value ${val}`)),
    filterNil(),
    take(3)
  )
  .subscribe(
    item => {
      console.log(`Rendering Item ${item}`);
    },
    err => console.log(err),
    () => console.log("Completed")
  );

console.log("\n\n\n\n");
console.log("With Filter Operator");
of(null, 20, 15, 10, 5, 6, 9)
  .pipe(
    filter(v => v !== null),
    take(3)
  )
  .subscribe(
    item => {
      console.log(`Rendering Item ${item}`);
    },
    err => console.log(err),
    () => console.log("Completed")
  );
```

**Expected behavior**
When I use the custom operator `filterNil`, I would expect that `tap` does print only the first 3 elements of the array, which is what happens when I use `filter` operator. On the contrary, when I use `filterNil` in the pipe, all the elements of the array get printed by the `tap` operator.

**Comment by Andrei0872** [comment]
You'll have to add these lines in order to solve this:

```typescript
/* ... */
next(value) {
  if (value !== undefined && value !== null) {
    subscriber.next(value);

    if (subscriber.closed) {
      this.unsubscribe();
    }
  }
},
/* ... */
```
Whenever you do `.subscribe()`, a `Subscriber` will be created. When using multiple operators, you'll end up having a **chain of Subscribers**.

For example:

```typescript
// S#n - Subscriber_n

of(/* ... */)
  .pipe(
    tap(val => console.log(`Tapped value ${val}`)), // S#4
    filterNil(),
    take(3) // S#2
  )
  .subscribe({ /* ... */ }) // S#1

// filterNil()
const subscription = source.subscribe() // S#3
```

The information flows from `S#4` to `S#1`.

The `subscriber` argument from `filterNil` will point to `S#2`.

But you're not _directly_ using `S#2`, you pass along information with the help of `S#3`:

```ts
const subscription = source.subscribe({ // This is S#3
  next(value) {
    if (value !== undefined && value !== null) {
      subscriber.next(value); // subscriber -> S#2
    }
  },
  error(error) {
    subscriber.error(error);
  },
  complete() {
    subscriber.complete();
  }
});
```

Due to `take(n)`, when 3 notifications are reached, it will unsubscribe. When that happens, `S#2` will become `closed`, but because `S#2` is not directly communicating with `S#4`, you'll have to manage this yourself, by checking for `closed` status after every `next notification`.

When you do `this.unsubscribe()`, this will point to `S#3`. But, at the same time, `S#3` is the parent of `S#4`, just like `S#1` is `S#2`'s parent. When a subscriber is unsubscribed, all of its children subscribers will be unsubscribed as well.

**Comment by EnricoPicci** [comment]
@Andrei0872 thanks for the clear explanation. It could be worth add this as an example to the documentation.
